core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -738,22 +748,34 @@ private void validateGetWithIndexResult(
             .build();
 
     LinkedHashMap<Key, TransactionResult> results = new LinkedHashMap<>(1);
-    TableMetadata tableMetadata = getTableMetadata(scanWithIndex);
-    originalResult.ifPresent(
-        r -> results.put(new Snapshot.Key(scanWithIndex, r, tableMetadata), r));
+    originalResult.ifPresent(r -> results.put(new Snapshot.Key(scanWithIndex, r, metadata), r));
 
     // Validate the result to check if there is no anti-dependency
     validateScanResults(storage, scanWithIndex, results, false);
   }
 
   private void validateGetResult(
-      DistributedStorage storage, Get get, Optional<TransactionResult> originalResult)
+      DistributedStorage storage,
+      Get get,
+      Optional<TransactionResult> originalResult,
+      TableMetadata metadata)
       throws ExecutionException, ValidationConflictException {
-    // Only get the tx_id column because we use only them to compare
-    get = Get.newBuilder(get).clearProjections().projection(Attribute.ID).build();
-
     // Check if a read record is not changed
-    Optional<TransactionResult> latestResult = storage.get(get).map(TransactionResult::new);
+    Optional<TransactionResult> latestResult =
+        storage
+            .get(ConsensusCommitUtils.prepareGetForStorage(get, metadata))
+            .map(TransactionResult::new);
+
+    if (!get.getConjunctions().isEmpty()) {
+      // Because we also get records whose before images match the conjunctions, we need to check if
+      // the current status of the records actually match the conjunctions.
+      latestResult =
+          latestResult.filter(
+              r ->
+                  ScalarDbUtils.columnsMatchAnyOfConjunctions(
+                      r.getColumns(), get.getConjunctions()));
+    }
```

Comment:
Because we also get records whose before images match the conjunctions, we need to check if the current status of the records actually match the conjunctions.